File: skills/manage-current-focus/scripts/knowledge_loom/contract.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any

import yaml

from .models import Finding, Vault
from .pathing import is_vault_relative_path, is_within_vault

CONTRACT_NAME = "KNOWLEDGE_VAULT.md"
VAULT_ID_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")


class ContractError(ValueError):
    pass


def split_frontmatter(text: str, *, source: str = "<text>") -> tuple[dict[str, Any], str]:
    lines = text.splitlines(keepends=True)
    if not lines or lines[0].strip() != "---":
        raise ContractError(f"{source}: missing opening YAML frontmatter delimiter")

    end = next((index for index, line in enumerate(lines[1:], start=1) if line.strip() == "---"), None)
    if end is None:
        raise ContractError(f"{source}: missing closing YAML frontmatter delimiter")

    raw = "".join(lines[1:end])
    try:
        data = yaml.safe_load(raw) or {}
    except yaml.YAMLError as exc:
        raise ContractError(f"{source}: invalid YAML frontmatter: {exc}") from exc
    if not isinstance(data, dict):
        raise ContractError(f"{source}: frontmatter must be a mapping")
    return data, "".join(lines[end + 1 :])


def load_vault(root: Path | str) -> Vault:
    root_path = Path(root).expanduser().resolve()
    contract_path = root_path / CONTRACT_NAME
    if not contract_path.is_file():
        raise ContractError(f"{contract_path}: contract not found")
    if not is_within_vault(root_path, contract_path):
        raise ContractError(f"{contract_path}: contract resolves outside the vault root")
    data, body = split_frontmatter(contract_path.read_text(encoding="utf-8"), source=str(contract_path))
    return Vault(root=root_path, contract_path=contract_path, contract=data, body=body)


def load_note_frontmatter(path: Path) -> dict[str, Any]:
    try:
        data, _ = split_frontmatter(path.read_text(encoding="utf-8"), source=str(path))
        return data
    except ContractError:
        return {}


def render_contract(data: dict[str, Any], body: str) -> str:
    yaml_text = yaml.safe_dump(
        data,
        sort_keys=False,
        allow_unicode=True,
        default_flow_style=False,
    ).rstrip()
    return f"---\n{yaml_text}\n---\n\n{body.strip()}\n"


def _mapping(contract: dict, key: str, findings: list[Finding]) -> dict:
    value = contract.get(key)
    if not isinstance(value, dict):
        findings.append(Finding("error", f"contract.{key}", f"`{key}` must be a mapping"))
        return {}
    return value


def _validate_path_values(
    values: object,
    *,
    field: str,
    findings: list[Finding],
    require_nonempty: bool = False,
) -> list[str]:
    if not isinstance(values, list) or (require_nonempty and not values) or any(
        not isinstance(item, str) for item in values
    ):
        qualifier = "non-empty " if require_nonempty else ""
        findings.append(Finding("error", f"contract.{field}", f"`{field}` must be a {qualifier}string list"))
        return []

    for value in values:
        if not is_vault_relative_path(value):
            findings.append(
                Finding(
                    "error",
                    "contract.path-boundary",
                    f"`{field}` path must stay within the vault root",
                    value,
                )
            )
    return values


def validate_contract_data(contract: dict[str, Any]) -> list[Finding]:
    findings: list[Finding] = []

    if contract.get("schema_version") != 1:
        findings.append(Finding("error", "contract.schema-version", "`schema_version` must equal 1"))

    vault_id = contract.get("vault_id")
    if not isinstance(vault_id, str) or not VAULT_ID_RE.fullmatch(vault_id):
        findings.append(Finding("error", "contract.vault-id", "`vault_id` must be stable kebab-case"))

    if not isinstance(contract.get("title"), str) or not contract.get("title", "").strip():
        findings.append(Finding("error", "contract.title", "`title` must be a non-empty string"))

    storage = _mapping(contract, "storage", findings)
    if storage and storage.get("type") != "local-markdown":
        findings.append(Finding("error", "contract.storage", "`storage.type` must be `local-markdown`"))

    subjects = _mapping(contract, "subjects", findings)
    mode = subjects.get("mode")
    values = subjects.get("values")
    if mode not in {"single", "multiple"}:
        findings.append(Finding("error", "contract.subject-mode", "`subjects.mode` must be `single` or `multiple`"))
    if not isinstance(values, list) or not values or any(not isinstance(item, str) for item in values):
        findings.append(Finding("error", "contract.subject-values", "`subjects.values` must be a non-empty string list"))
        values = []
    elif len(set(values)) != len(values):
        findings.append(Finding("error", "contract.subject-values", "`subjects.values` must be unique"))
    if mode == "single":
        if len(values) != 1:
            findings.append(Finding("error", "contract.single-subject", "single-subject vaults require exactly one value"))
        if values and subjects.get("default") != values[0]:
            findings.append(Finding("error", "contract.subject-default", "single-subject default must equal its only value"))

    write = _mapping(contract, "write", findings)
    if write and write.get("policy") not in {"explicit-only", "proactive-durable-capture"}:
        findings.append(Finding("error", "contract.write-policy", "unsupported `write.policy`"))
    if write and write.get("current_state_policy") not in {
        "explicit-only",
        "maintain-after-material-change",
    }:
        findings.append(Finding("error", "contract.current-state-policy", "unsupported current-state policy"))

    history = _mapping(contract, "history", findings)
    if history and history.get("type") not in {"git", "none"}:
        findings.append(Finding("error", "contract.history", "`history.type` must be `git` or `none`"))

    sync = _mapping(contract, "sync", findings)
    if sync and sync.get("mode") not in {"none", "git-remote-push", "lifecycle-hook"}:
        findings.append(Finding("error", "contract.sync", "unsupported `sync.mode`"))
    if sync.get("mode") == "lifecycle-hook" and not sync.get("adapter"):
        findings.append(Finding("error", "contract.sync-adapter", "lifecycle sync requires `adapter`"))

    backup = _mapping(contract, "backup", findings)
    if backup and backup.get("mode") not in {"none", "lifecycle-hook"}:
        findings.append(Finding("error", "contract.backup", "unsupported `backup.mode`"))
    if backup.get("mode") == "lifecycle-hook" and not backup.get("adapter"):
        findings.append(Finding("error", "contract.backup-adapter", "lifecycle backup requires `adapter`"))

    _validate_path_values(contract.get("instruction_roots"), field="instruction_roots", findings=findings)

    navigation = _mapping(contract, "navigation", findings)
    _validate_path_values(
        navigation.get("entrypoints"),
        field="navigation.entrypoints",
        findings=findings,
        require_nonempty=True,
    )

    profiles = contract.get("metadata_profiles", {})
    if not isinstance(profiles, dict):
        findings.append(Finding("error", "contract.metadata-profiles", "`metadata_profiles` must be a mapping"))
    else:
        for name, profile in profiles.items():
            if not isinstance(profile, dict):
                findings.append(Finding("error", "contract.metadata-profile", f"profile `{name}` must be a mapping"))
                continue
            paths = profile.get("paths")
            if not isinstance(paths, list) or not paths or any(not isinstance(item, str) for item in paths):
                findings.append(Finding("error", "contract.metadata-paths", f"profile `{name}` requires string paths"))
            else:
                for path in paths:
                    if not is_vault_relative_path(path):
                        findings.append(
                            Finding(
                                "error",
                                "contract.path-boundary",
                                f"metadata profile `{name}` path must stay within the vault root",
                                path,
                            )
                        )
            if not isinstance(profile.get("required"), list):
                findings.append(Finding("error", "contract.metadata-required", f"profile `{name}` requires a field list"))
            if profile.get("severity", "error") not in {"error", "warning"}:
                findings.append(Finding("error", "contract.metadata-severity", f"profile `{name}` has invalid severity"))

    focus_views = contract.get("focus_views", {})
    if not isinstance(focus_views, dict):
        findings.append(Finding("error", "contract.focus-views", "`focus_views` must be a mapping"))
    else:
        subject_values = set(values)
        for name, view in focus_views.items():
            if not isinstance(view, dict):
                findings.append(Finding("error", "contract.focus-view", f"focus view `{name}` must be a mapping"))
                continue
            view_path = view.get("path")
            if not isinstance(view_path, str):
                findings.append(Finding("error", "contract.focus-path", f"focus view `{name}` requires `path`"))
            elif not is_vault_relative_path(view_path):
                findings.append(
                    Finding(
                        "error",
                        "contract.path-boundary",
                        f"focus view `{name}` path must stay within the vault root",
                        view_path,
                    )
                )
            if view.get("subject") not in subject_values:
                findings.append(Finding("error", "contract.focus-subject", f"focus view `{name}` has unknown subject"))
            for key in ("max_active", "max_top"):
                if not isinstance(view.get(key), int) or view[key] < 1:
                    findings.append(Finding("error", f"contract.focus-{key}", f"focus view `{name}` requires positive `{key}`"))
            if isinstance(view.get("max_active"), int) and isinstance(view.get("max_top"), int):
                if view["max_active"] > view["max_top"]:
                    findings.append(Finding("error", "contract.focus-limits", f"focus view `{name}` has max_active > max_top"))

    privacy = contract.get("privacy", {})
    if not isinstance(privacy, dict):
        findings.append(Finding("error", "contract.privacy", "`privacy.never_track` must be a string list"))
    else:
        _validate_path_values(privacy.get("never_track", []), field="privacy.never_track", findings=findings)

    return findings

File: src/knowledge-loom/contract.mjs
import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";

import { ContractError } from "./errors.mjs";
import { canonicalPath, isVaultRelativePath, isWithin } from "./pathing.mjs";

export { ContractError } from "./errors.mjs";

export const CONTRACT_NAME = "KNOWLEDGE_VAULT.md";
const VAULT_ID_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

export function finding(severity, code, message, findingPath = null) {
  return { severity, code, message, path: findingPath };
}

export function splitFrontmatter(text, { source = "<text>" } = {}) {
  const lines = text.match(/.*(?:\r?\n|$)/g)?.filter((line) => line.length) ?? [];
  if (!lines.length || lines[0].trim() !== "---") {
    throw new ContractError(`${source}: missing opening YAML frontmatter delimiter`);
  }
  const end = lines.findIndex((line, index) => index > 0 && line.trim() === "---");
  if (end < 0) throw new ContractError(`${source}: missing closing YAML frontmatter delimiter`);

  const raw = lines.slice(1, end).join("");
  let data;
  try {
    data = YAML.parse(raw) ?? {};
  } catch (error) {
    throw new ContractError(`${source}: invalid YAML frontmatter: ${error.message}`, { cause: error });
  }
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    throw new ContractError(`${source}: frontmatter must be a mapping`);
  }
  return [data, lines.slice(end + 1).join("")];
}

export function loadVault(root) {
  const rootPath = canonicalPath(root);
  const contractPath = path.join(rootPath, CONTRACT_NAME);
  if (!fs.statSync(contractPath, { throwIfNoEntry: false })?.isFile()) {
    throw new ContractError(`${contractPath}: contract not found`);
  }
  if (!isWithin(rootPath, contractPath)) {
    throw new ContractError(`${contractPath}: contract resolves outside the vault root`);
  }
  const [contract, body] = splitFrontmatter(fs.readFileSync(contractPath, "utf8"), { source: contractPath });
  return { root: rootPath, contractPath, contract, body };
}

export function loadNoteFrontmatter(notePath) {
  try {
    return splitFrontmatter(fs.readFileSync(notePath, "utf8"), { source: notePath })[0];
  } catch (error) {
    if (error instanceof ContractError) return {};
    throw error;
  }
}

export function renderContract(data, body) {
  const yamlText = YAML.stringify(data, { lineWidth: 0 }).trimEnd();
  return `---\n${yamlText}\n---\n\n${body.trim()}\n`;
}

function mapping(contract, key, findings) {
  const value = contract[key];
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    findings.push(finding("error", `contract.${key}`, `\`${key}\` must be a mapping`));
    return {};
  }
  return value;
}

function validatePathValue(value, description, findings) {
  if (!isVaultRelativePath(value)) {
    findings.push(finding("error", "contract.path-boundary", `${description} must stay within the vault root`, value));
  }
}

function validatePathValues(values, { field, findings, requireNonempty = false }) {
  if (!Array.isArray(values) || (requireNonempty && values.length === 0) || values.some((item) => typeof item !== "string")) {
    const qualifier = requireNonempty ? "non-empty " : "";
    findings.push(finding("error", `contract.${field}`, `\`${field}\` must be a ${qualifier}string list`));
    return [];
  }
  for (const value of values) validatePathValue(value, `\`${field}\` path`, findings);
  return values;
}

export function validateContractData(contract) {
  const findings = [];
  if (contract.schema_version !== 1) findings.push(finding("error", "contract.schema-version", "`schema_version` must equal 1"));

  const vaultId = contract.vault_id;
  if (typeof vaultId !== "string" || !VAULT_ID_RE.test(vaultId)) {
    findings.push(finding("error", "contract.vault-id", "`vault_id` must be stable kebab-case"));
  }
  if (typeof contract.title !== "string" || !contract.title.trim()) {
    findings.push(finding("error", "contract.title", "`title` must be a non-empty string"));
  }

  const storage = mapping(contract, "storage", findings);
  if (Object.keys(storage).length && storage.type !== "local-markdown") {
    findings.push(finding("error", "contract.storage", "`storage.type` must be `local-markdown`"));
  }

  const subjects = mapping(contract, "subjects", findings);
  const mode = subjects.mode;
  let values = subjects.values;
  if (!new Set(["single", "multiple"]).has(mode)) {
    findings.push(finding("error", "contract.subject-mode", "`subjects.mode` must be `single` or `multiple`"));
  }
  if (!Array.isArray(values) || !values.length || values.some((item) => typeof item !== "string")) {
    findings.push(finding("error", "contract.subject-values", "`subjects.values` must be a non-empty string list"));
    values = [];
  } else if (new Set(values).size !== values.length) {
    findings.push(finding("error", "contract.subject-values", "`subjects.values` must be unique"));
  }
  if (mode === "single") {
    if (values.length !== 1) findings.push(finding("error", "contract.single-subject", "single-subject vaults require exactly one value"));
    if (values.length && subjects.default !== values[0]) {
      findings.push(finding("error", "contract.subject-default", "single-subject default must equal its only value"));
    }
  }

  const write = mapping(contract, "write", findings);
  if (Object.keys(write).length && !new Set(["explicit-only", "proactive-durable-capture"]).has(write.policy)) {
    findings.push(finding("error", "contract.write-policy", "unsupported `write.policy`"));
  }
  if (Object.keys(write).length && !new Set(["explicit-only", "maintain-after-material-change"]).has(write.current_state_policy)) {
    findings.push(finding("error", "contract.current-state-policy", "unsupported current-state policy"));
  }

  const history = mapping(contract, "history", findings);
  if (Object.keys(history).length && !new Set(["git", "none"]).has(history.type)) {
    findings.push(finding("error", "contract.history", "`history.type` must be `git` or `none`"));
  }

  const sync = mapping(contract, "sync", findings);
  if (Object.keys(sync).length && !new Set(["none", "git-remote-push", "lifecycle-hook"]).has(sync.mode)) {
    findings.push(finding("error", "contract.sync", "unsupported `sync.mode`"));
  }
  if (sync.mode === "lifecycle-hook" && !sync.adapter) {
    findings.push(finding("error", "contract.sync-adapter", "lifecycle sync requires `adapter`"));
  }

  const backup = mapping(contract, "backup", findings);
  if (Object.keys(backup).length && !new Set(["none", "lifecycle-hook"]).has(backup.mode)) {
    findings.push(finding("error", "contract.backup", "unsupported `backup.mode`"));
  }
  if (backup.mode === "lifecycle-hook" && !backup.adapter) {
    findings.push(finding("error", "contract.backup-adapter", "lifecycle backup requires `adapter`"));
  }

  if (Object.hasOwn(contract, "content_checks")) {
    const contentChecks = mapping(contract, "content_checks", findings);
    if (
      contract.content_checks
      && typeof contract.content_checks === "object"
      && !Array.isArray(contract.content_checks)
      && (typeof contentChecks.adapter !== "string" || !VAULT_ID_RE.test(contentChecks.adapter))
    ) {
      findings.push(
        finding(
          "error",
          "contract.content-check-adapter",
          "`content_checks.adapter` must be a kebab-case ID",
        ),
      );
    }
  }

  validatePathValues(contract.instruction_roots, { field: "instruction_roots", findings });
  const navigation = mapping(contract, "navigation", findings);
  validatePathValues(navigation.entrypoints, { field: "navigation.entrypoints", findings, requireNonempty: true });

  const profiles = contract.metadata_profiles ?? {};
  if (!profiles || typeof profiles !== "object" || Array.isArray(profiles)) {
    findings.push(finding("error", "contract.metadata-profiles", "`metadata_profiles` must be a mapping"));
  } else {
    for (const [name, profile] of Object.entries(profiles)) {
      if (!profile || typeof profile !== "object" || Array.isArray(profile)) {
        findings.push(finding("error", "contract.metadata-profile", `profile \`${name}\` must be a mapping`));
        continue;
      }
      if (!Array.isArray(profile.paths) || !profile.paths.length || profile.paths.some((item) => typeof item !== "string")) {
        findings.push(finding("error", "contract.metadata-paths", `profile \`${name}\` requires string paths`));
      } else {
        for (const profilePath of profile.paths) validatePathValue(profilePath, `metadata profile \`${name}\` path`, findings);
      }
      if (!Array.isArray(profile.required)) findings.push(finding("error", "contract.metadata-required", `profile \`${name}\` requires a field list`));
      if (!new Set(["error", "warning"]).has(profile.severity ?? "error")) {
        findings.push(finding("error", "contract.metadata-severity", `profile \`${name}\` has invalid severity`));
      }
    }
  }

  const focusViews = contract.focus_views ?? {};
  if (!focusViews || typeof focusViews !== "object" || Array.isArray(focusViews)) {
    findings.push(finding("error", "contract.focus-views", "`focus_views` must be a mapping"));
  } else {
    const subjectValues = new Set(values);
    for (const [name, view] of Object.entries(focusViews)) {
      if (!view || typeof view !== "object" || Array.isArray(view)) {
        findings.push(finding("error", "contract.focus-view", `focus view \`${name}\` must be a mapping`));
        continue;
      }
      if (typeof view.path !== "string") findings.push(finding("error", "contract.focus-path", `focus view \`${name}\` requires \`path\``));
      else validatePathValue(view.path, `focus view \`${name}\` path`, findings);
      if (!subjectValues.has(view.subject)) findings.push(finding("error", "contract.focus-subject", `focus view \`${name}\` has unknown subject`));
      for (const key of ["max_active", "max_top"]) {
        if (!Number.isInteger(view[key]) || view[key] < 1) findings.push(finding("error", `contract.focus-${key}`, `focus view \`${name}\` requires positive \`${key}\``));
      }
      if (Number.isInteger(view.max_active) && Number.isInteger(view.max_top) && view.max_active > view.max_top) {
        findings.push(finding("error", "contract.focus-limits", `focus view \`${name}\` has max_active > max_top`));
      }
    }
  }

  const privacy = contract.privacy ?? {};
  if (!privacy || typeof privacy !== "object" || Array.isArray(privacy)) {
    findings.push(finding("error", "contract.privacy", "`privacy.never_track` must be a string list"));
  } else {
    validatePathValues(privacy.never_track ?? [], { field: "privacy.never_track", findings });
  }
  return findings;
}
